File: swift/Sources/SnapAPI/SnapAPI.swift
import Foundation

/// SnapAPI SDK for iOS/macOS
///
/// Example usage:
/// ```swift
/// let client = SnapAPI(apiKey: "sk_live_xxx")
///
/// // Capture a screenshot
/// let screenshot = try await client.screenshot(
///     ScreenshotOptions(url: "https://example.com")
/// )
///
/// // Save to file
/// try screenshot.write(to: URL(fileURLWithPath: "screenshot.png"))
/// ```
public actor SnapAPI {
    /// SDK version
    public static let version = "1.3.0"

    /// Default base URL
    public static let defaultBaseURL = "https://api.snapapi.pics"

    /// Default timeout in seconds
    public static let defaultTimeout: TimeInterval = 60

    private let apiKey: String
    private let baseURL: String
    private let session: URLSession

    /// Initialize the SnapAPI client.
    ///
    /// - Parameters:
    ///   - apiKey: Your SnapAPI API key
    ///   - baseURL: Optional custom base URL
    ///   - timeout: Request timeout in seconds (default: 60)
    public init(
        apiKey: String,
        baseURL: String = defaultBaseURL,
        timeout: TimeInterval = defaultTimeout
    ) {
        precondition(!apiKey.isEmpty, "API key is required")

        self.apiKey = apiKey
        self.baseURL = baseURL.trimmingCharacters(in: CharacterSet(charactersIn: "/"))

        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = timeout
        config.timeoutIntervalForResource = timeout
        self.session = URLSession(configuration: config)
    }

    // MARK: - Ping

    /// Check API health.
    ///
    /// - Returns: Ping result with status and timestamp
    public func ping() async throws -> PingResult {
        let data: Data = try await doRequest("GET", path: "/v1/ping", body: nil as Empty?)
        return try JSONDecoder().decode(PingResult.self, from: data)
    }

    // MARK: - Screenshot Methods

    /// Capture a screenshot of the specified URL or HTML content.
    ///
    /// - Parameter options: Screenshot options
    /// - Returns: Raw image data
    public func screenshot(_ options: ScreenshotOptions) async throws -> Data {
        guard options.url != nil || options.html != nil || options.markdown != nil else {
            throw SnapAPIError(
                code: "INVALID_PARAMS",
                message: "Either url, html, or markdown is required",
                statusCode: 400
            )
        }

        return try await doRequest("POST", path: "/v1/screenshot", body: options)
    }

    /// Capture a screenshot and return metadata.
    ///
    /// - Parameter options: Screenshot options (responseType will be set to json)
    /// - Returns: Screenshot result with metadata
    public func screenshotWithMetadata(_ options: ScreenshotOptions) async throws -> ScreenshotResult {
        var opts = options
        opts.responseType = "json"

        let data = try await screenshot(opts)
        return try JSONDecoder().decode(ScreenshotResult.self, from: data)
    }

    /// Capture a screenshot from HTML content.
    ///
    /// - Parameters:
    ///   - html: HTML content to render
    ///   - options: Additional screenshot options
    /// - Returns: Raw image data
    public func screenshotFromHtml(_ html: String, options: ScreenshotOptions? = nil) async throws -> Data {
        var opts = options ?? ScreenshotOptions()
        opts.html = html
        opts.url = nil
        return try await screenshot(opts)
    }

    /// Capture a screenshot from Markdown content.
    ///
    /// - Parameters:
    ///   - markdown: Markdown content to render
    ///   - options: Additional screenshot options
    /// - Returns: Raw image data
    public func screenshotFromMarkdown(_ markdown: String, options: ScreenshotOptions? = nil) async throws -> Data {
        var opts = options ?? ScreenshotOptions()
        opts.markdown = markdown
        opts.url = nil
        opts.html = nil
        return try await screenshot(opts)
    }

    /// Capture a screenshot using a device preset.
    ///
    /// - Parameters:
    ///   - url: URL to capture
    ///   - device: Device preset
    ///   - options: Additional screenshot options
    /// - Returns: Raw image data
    public func screenshotDevice(
        url: String,
        device: DevicePreset,
        options: ScreenshotOptions? = nil
    ) async throws -> Data {
        var opts = options ?? ScreenshotOptions()
        opts.url = url
        opts.device = device.rawValue
        return try await screenshot(opts)
    }

    // MARK: - Async Screenshot

    /// Submit an async screenshot job.
    ///
    /// - Parameter options: Screenshot options
    /// - Returns: Async job result with jobId and statusUrl
    public func screenshotAsync(_ options: ScreenshotOptions) async throws -> AsyncJobResult {
        guard options.url != nil || options.html != nil || options.markdown != nil else {
            throw SnapAPIError(
                code: "INVALID_PARAMS",
                message: "Either url, html, or markdown is required",
                statusCode: 400
            )
        }

        var opts = options
        opts.async = true
        opts.responseType = "json"

        let data = try await doRequest("POST", path: "/v1/screenshot", body: opts)
        return try JSONDecoder().decode(AsyncJobResult.self, from: data)
    }

    /// Poll async screenshot job status.
    ///
    /// - Parameter jobId: The async job ID
    /// - Returns: Async status result
    public func getAsyncStatus(_ jobId: String) async throws -> AsyncStatusResult {
        let data: Data = try await doRequest("GET", path: "/v1/screenshot/async/\(jobId)", body: nil as Empty?)
        return try JSONDecoder().decode(AsyncStatusResult.self, from: data)
    }

    /// Submit an async screenshot and poll until completion.
    ///
    /// - Parameters:
    ///   - options: Screenshot options
    ///   - pollInterval: Seconds between polls (default: 1)
    ///   - maxAttempts: Maximum poll attempts (default: 60)
    /// - Returns: Completed async status result
    public func screenshotAsyncAndWait(
        _ options: ScreenshotOptions,
        pollInterval: TimeInterval = 1,
        maxAttempts: Int = 60
    ) async throws -> AsyncStatusResult {
        let job = try await screenshotAsync(options)

        for _ in 0..<maxAttempts {
            try await Task.sleep(nanoseconds: UInt64(pollInterval * 1_000_000_000))
            let status = try await getAsyncStatus(job.jobId)
            if status.status == "completed" || status.status == "failed" {
                return status
            }
        }

        throw SnapAPIError(
            code: "TIMEOUT",
            message: "Async screenshot timed out after \(maxAttempts) attempts",
            statusCode: 408
        )
    }

    // MARK: - PDF Methods

    /// Generate a PDF using the dedicated /v1/pdf endpoint.
    ///
    /// - Parameter options: Screenshot options with PDF-specific settings
    /// - Returns: Raw PDF data
    public func pdf(_ options: ScreenshotOptions) async throws -> Data {
        guard options.url != nil || options.html != nil || options.markdown != nil else {
            throw SnapAPIError(
                code: "INVALID_PARAMS",
                message: "Either url, html, or markdown is required",
                statusCode: 400
            )
        }

        return try await doRequest("POST", path: "/v1/pdf", body: options)
    }

    /// Generate a PDF from HTML content.
    ///
    /// - Parameters:
    ///   - html: HTML content to render
    ///   - pdfOptions: PDF generation options
    /// - Returns: Raw PDF data
    public func pdfFromHtml(_ html: String, pdfOptions: PdfOptions? = nil) async throws -> Data {
        return try await pdf(ScreenshotOptions(
            html: html,
            pdfOptions: pdfOptions
        ))
    }

    // MARK: - Video Methods

    /// Capture a video of a webpage with optional scroll animation.
    ///
    /// - Parameter options: Video options
    /// - Returns: Raw video data
    public func video(_ options: VideoOptions) async throws -> Data {
        return try await doRequest("POST", path: "/v1/video", body: options)
    }

    /// Capture a video and return structured result with metadata.
    ///
    /// - Parameter options: Video options
    /// - Returns: Video result with metadata
    public func videoWithResult(_ options: VideoOptions) async throws -> VideoResult {
        var opts = options
        opts.responseType = "json"
        let data = try await doRequest("POST", path: "/v1/video", body: opts)
        return try JSONDecoder().decode(VideoResult.self, from: data)
    }

    // MARK: - Batch Methods

    /// Capture screenshots of multiple URLs.
    ///
    /// - Parameter options: Batch options
    /// - Returns: Batch result with job ID
    public func batch(_ options: BatchOptions) async throws -> BatchResult {
        guard !options.urls.isEmpty else {
            throw SnapAPIError(
                code: "INVALID_PARAMS",
                message: "URLs are required",
                statusCode: 400
            )
        }

        let data = try await doRequest("POST", path: "/v1/screenshot/batch", body: options)
        return try JSONDecoder().decode(BatchResult.self, from: data)
    }

    /// Check the status of a batch job.
    ///
    /// - Parameter jobId: The batch job ID
    /// - Returns: Batch status with results if completed
    public func getBatchStatus(_ jobId: String) async throws -> BatchStatus {
        let data: Data = try await doRequest("GET", path: "/v1/screenshot/batch/\(jobId)", body: nil as Empty?)
        return try JSONDecoder().decode(BatchStatus.self, from: data)
    }

    /// Submit a batch and poll until completion.
    ///
    /// - Parameters:
    ///   - options: Batch options
    ///   - pollInterval: Seconds between polls (default: 2)
    ///   - maxAttempts: Maximum poll attempts (default: 60)
    /// - Returns: Completed batch status
    public func batchAndWait(
        _ options: BatchOptions,
        pollInterval: TimeInterval = 2,
        maxAttempts: Int = 60
    ) async throws -> BatchStatus {
        let result = try await batch(options)

        for _ in 0..<maxAttempts {
            try await Task.sleep(nanoseconds: UInt64(pollInterval * 1_000_000_000))
            let status = try await getBatchStatus(result.jobId)
            if status.status == "completed" || status.status == "failed" {
                return status
            }
        }

        throw SnapAPIError(
            code: "TIMEOUT",
            message: "Batch job timed out after \(maxAttempts) attempts",
            statusCode: 408
        )
    }

    // MARK: - Extract API

    /// Extract content from a webpage.
    public func extract(_ options: ExtractOptions) async throws -> Data {
        guard !options.url.isEmpty else {
            throw SnapAPIError(code: "INVALID_PARAMS", message: "URL is required", statusCode: 400)
        }
        return try await doRequest("POST", path: "/v1/extract", body: options)
    }

    /// Extract markdown from a webpage.
    public func extractMarkdown(_ url: String) async throws -> Data {
        return try await extract(ExtractOptions(url: url, type: .markdown))
    }

    /// Extract article content from a webpage.
    public func extractArticle(_ url: String) async throws -> Data {
        return try await extract(ExtractOptions(url: url, type: .article))
    }

    /// Extract structured data for LLM/RAG workflows.
    public func extractStructured(_ url: String) async throws -> Data {
        return try await extract(ExtractOptions(url: url, type: .structured))
    }

    /// Extract plain text from a webpage.
    public func extractText(_ url: String) async throws -> Data {
        return try await extract(ExtractOptions(url: url, type: .text))
    }

    /// Extract all links from a webpage.
    public func extractLinks(_ url: String) async throws -> Data {
        return try await extract(ExtractOptions(url: url, type: .links))
    }

    /// Extract all images from a webpage.
    public func extractImages(_ url: String) async throws -> Data {
        return try await extract(ExtractOptions(url: url, type: .images))
    }

    /// Extract page metadata from a webpage.
    public func extractMetadata(_ url: String) async throws -> Data {
        return try await extract(ExtractOptions(url: url, type: .metadata))
    }

    // MARK: - Analyze API

    /// Analyze a webpage using AI (BYOK - Bring Your Own Key).
    public func analyze(_ options: AnalyzeOptions) async throws -> Data {
        guard !options.url.isEmpty else {
            throw SnapAPIError(code: "INVALID_PARAMS", message: "URL is required", statusCode: 400)
        }
        guard !options.prompt.isEmpty else {
            throw SnapAPIError(code: "INVALID_PARAMS", message: "Prompt is required", statusCode: 400)
        }
        guard !options.apiKey.isEmpty else {
            throw SnapAPIError(code: "INVALID_PARAMS", message: "LLM API key is required", statusCode: 400)
        }
        return try await doRequest("POST", path: "/v1/analyze", body: options)
    }

    // MARK: - Info Methods

    /// Get available device presets.
    public func getDevices() async throws -> DevicesResult {
        let data: Data = try await doRequest("GET", path: "/v1/devices", body: nil as Empty?)
        return try JSONDecoder().decode(DevicesResult.self, from: data)
    }

    /// Get API capabilities and features.
    public func getCapabilities() async throws -> CapabilitiesResult {
        let data: Data = try await doRequest("GET", path: "/v1/capabilities", body: nil as Empty?)
        return try JSONDecoder().decode(CapabilitiesResult.self, from: data)
    }

    /// Get API usage statistics.
    public func getUsage() async throws -> UsageResult {
        let data: Data = try await doRequest("GET", path: "/v1/usage", body: nil as Empty?)
        return try JSONDecoder().decode(UsageResult.self, from: data)
    }

    // MARK: - Private Methods

    private func doRequest<T: Encodable>(_ method: String, path: String, body: T?) async throws -> Data {
        guard let url = URL(string: "\(baseURL)\(path)") else {
            throw SnapAPIError(
                code: "INVALID_URL",
                message: "Invalid URL",
                statusCode: 400
            )
        }

        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue(apiKey, forHTTPHeaderField: "X-Api-Key")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("snapapi-swift/\(Self.version)", forHTTPHeaderField: "User-Agent")

        if let body = body {
            let encoder = JSONEncoder()
            encoder.keyEncodingStrategy = .useDefaultKeys
            request.httpBody = try encoder.encode(body)
        }

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw SnapAPIError(
                code: "CONNECTION_ERROR",
                message: "Invalid response",
                statusCode: 0
            )
        }

        if httpResponse.statusCode >= 400 {
            try handleError(data: data, statusCode: httpResponse.statusCode)
        }

        return data
    }

    private func handleError(data: Data, statusCode: Int) throws -> Never {
        do {
            let errorResponse = try JSONDecoder().decode(FlatErrorResponse.self, from: data)
            let code = errorResponse.error
                .uppercased()
                .replacingOccurrences(of: " ", with: "_")
            throw SnapAPIError(
                code: code,
                message: errorResponse.message,
                statusCode: statusCode,
                details: errorResponse.details
            )
        } catch let error as SnapAPIError {
            throw error
        } catch {
            throw SnapAPIError(
                code: "HTTP_ERROR",
                message: "HTTP \(statusCode)",
                statusCode: statusCode
            )
        }
    }
}

// MARK: - Empty Type for GET requests

private struct Empty: Encodable {}

// MARK: - SnapAPIError

/// Error thrown by SnapAPI operations.
public struct SnapAPIError: Error, LocalizedError, Sendable {
    /// Error code
    public let code: String

    /// Error message
    public let message: String

    /// HTTP status code
    public let statusCode: Int

    /// Additional error details
    public let details: [AnyCodable]?

    public init(
        code: String,
        message: String,
        statusCode: Int,
        details: [AnyCodable]? = nil
    ) {
        self.code = code
        self.message = message
        self.statusCode = statusCode
        self.details = details
    }

    public var errorDescription: String? {
        "[\(code)] \(message) (HTTP \(statusCode))"
    }

    /// Check if the error is retryable.
    public var isRetryable: Bool {
        code == "RATE_LIMITED" || code == "TIMEOUT" || statusCode >= 500
    }
}

// MARK: - Convenience Extensions

#if canImport(UIKit)
import UIKit

extension ScreenshotResult {
    /// Convert the screenshot data to UIImage
    public var uiImage: UIImage? {
        imageData.flatMap { UIImage(data: $0) }
    }

    /// Convert the thumbnail data to UIImage
    public var thumbnailUIImage: UIImage? {
        thumbnailData.flatMap { UIImage(data: $0) }
    }
}
#endif

#if canImport(AppKit)
import AppKit

extension ScreenshotResult {
    /// Convert the screenshot data to NSImage
    public var nsImage: NSImage? {
        imageData.flatMap { NSImage(data: $0) }
    }

    /// Convert the thumbnail data to NSImage
    public var thumbnailNSImage: NSImage? {
        thumbnailData.flatMap { NSImage(data: $0) }
    }
}
#endif

File: kotlin/src/main/kotlin/dev/snapapi/SnapAPI.kt
package dev.snapapi

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import java.util.Base64

/**
 * SnapAPI SDK for Android/Kotlin
 *
 * Example usage:
 * ```kotlin
 * val client = SnapAPI("sk_live_xxx")
 *
 * // Capture a screenshot
 * val screenshot = client.screenshot(
 *     ScreenshotOptions(url = "https://example.com")
 * )
 *
 * // Save to file
 * File("screenshot.png").writeBytes(screenshot)
 * ```
 */
class SnapAPI(
    private val apiKey: String,
    private val baseUrl: String = DEFAULT_BASE_URL,
    private val timeout: Int = DEFAULT_TIMEOUT
) {
    companion object {
        const val DEFAULT_BASE_URL = "https://api.snapapi.pics"
        const val DEFAULT_TIMEOUT = 60000
        const val VERSION = "1.1.0"
        private const val USER_AGENT = "snapapi-kotlin/$VERSION"
    }

    private val json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = false
        explicitNulls = false
    }

    init {
        require(apiKey.isNotBlank()) { "API key is required" }
    }

    /**
     * Capture a screenshot of the specified URL or HTML content.
     *
     * @param options Screenshot options
     * @return Raw image bytes
     */
    suspend fun screenshot(options: ScreenshotOptions): ByteArray {
        require(options.url != null || options.html != null) {
            "Either url or html is required"
        }

        return doRequest("POST", "/v1/screenshot", options)
    }

    /**
     * Capture a screenshot and return metadata.
     *
     * @param options Screenshot options (responseType will be set to json)
     * @return Screenshot result with metadata
     */
    suspend fun screenshotWithMetadata(options: ScreenshotOptions): ScreenshotResult {
        val opts = options.copy(responseType = "json")
        val response = doRequest("POST", "/v1/screenshot", opts)
        return json.decodeFromString(response.decodeToString())
    }

    /**
     * Capture a screenshot from HTML content.
     *
     * @param html HTML content to render
     * @param options Additional screenshot options
     * @return Raw image bytes
     */
    suspend fun screenshotFromHtml(html: String, options: ScreenshotOptions? = null): ByteArray {
        val opts = options?.copy(html = html, url = null)
            ?: ScreenshotOptions(html = html)
        return screenshot(opts)
    }

    /**
     * Capture a screenshot using a device preset.
     *
     * @param url URL to capture
     * @param device Device preset name (use DevicePresets constants)
     * @param options Additional screenshot options
     * @return Raw image bytes
     */
    suspend fun screenshotDevice(
        url: String,
        device: String,
        options: ScreenshotOptions? = null
    ): ByteArray {
        val opts = options?.copy(url = url, device = device)
            ?: ScreenshotOptions(url = url, device = device)
        return screenshot(opts)
    }

    /**
     * Generate a PDF from a URL or HTML content.
     *
     * @param options Screenshot options (format will be set to pdf)
     * @return Raw PDF bytes
     */
    suspend fun pdf(options: ScreenshotOptions): ByteArray {
        require(options.url != null || options.html != null) {
            "Either url or html is required"
        }

        val opts = options.copy(format = "pdf", responseType = "binary")
        return doRequest("POST", "/v1/screenshot", opts)
    }

    /**
     * Generate a PDF from HTML content.
     *
     * @param html HTML content to render
     * @param pdfOptions PDF generation options
     * @return Raw PDF bytes
     */
    suspend fun pdfFromHtml(html: String, pdfOptions: PdfOptions? = null): ByteArray {
        return pdf(ScreenshotOptions(
            html = html,
            format = "pdf",
            pdfOptions = pdfOptions
        ))
    }

    /**
     * Capture a video of a webpage with optional scroll animation.
     *
     * @param options Video options
     * @return Raw video bytes
     */
    suspend fun video(options: VideoOptions): ByteArray {
        return doRequest("POST", "/v1/video", options)
    }

    /**
     * Capture a video and return structured result with metadata.
     *
     * @param options Video options
     * @return Video result with metadata
     */
    suspend fun videoWithResult(options: VideoOptions): VideoResult {
        val opts = options.copy(responseType = "json")
        val response = doRequest("POST", "/v1/video", opts)
        return json.decodeFromString(response.decodeToString())
    }

    /**
     * Capture screenshots of multiple URLs.
     *
     * @param options Batch options
     * @return Batch result with job ID
     */
    suspend fun batch(options: BatchOptions): BatchResult {
        require(options.urls.isNotEmpty()) { "URLs are required" }

        val response = doRequest("POST", "/v1/screenshot/batch", options)
        return json.decodeFromString(response.decodeToString())
    }

    /**
     * Check the status of a batch job.
     *
     * @param jobId The batch job ID
     * @return Batch status with results if completed
     */
    suspend fun getBatchStatus(jobId: String): BatchStatus {
        val response = doRequest("GET", "/v1/screenshot/batch/$jobId", null)
        return json.decodeFromString(response.decodeToString())
    }

    /**
     * Get available device presets.
     *
     * @return Devices result with presets grouped by category
     */
    suspend fun getDevices(): DevicesResult {
        val response = doRequest("GET", "/v1/devices", null)
        return json.decodeFromString(response.decodeToString())
    }

    /**
     * Get API capabilities and features.
     *
     * @return Capabilities result
     */
    suspend fun getCapabilities(): CapabilitiesResult {
        val response = doRequest("GET", "/v1/capabilities", null)
        return json.decodeFromString(response.decodeToString())
    }

    /**
     * Get API usage statistics.
     *
     * @return Usage result with limits and remaining quota
     */
    suspend fun getUsage(): UsageResult {
        val response = doRequest("GET", "/v1/usage", null)
        return json.decodeFromString(response.decodeToString())
    }

    /**
     * Perform an HTTP request to the API.
     */

    // Extract API
    
    /**
     * Extract content from a webpage.
     */
    suspend fun extract(options: ExtractOptions): ExtractResult {
        require(options.url.isNotBlank()) { "URL is required" }
        
        val response = doRequest("POST", "/v1/extract", options)
        return json.decodeFromString(response.decodeToString())
    }
    
    /**
     * Extract markdown from a webpage.
     */
    suspend fun extractMarkdown(url: String): ExtractResult {
        return extract(ExtractOptions(url = url, type = ExtractType.markdown))
    }
    
    /**
     * Extract article content from a webpage.
     */
    suspend fun extractArticle(url: String): ExtractResult {
        return extract(ExtractOptions(url = url, type = ExtractType.article))
    }
    
    /**
     * Extract structured data for LLM/RAG workflows.
     */
    suspend fun extractStructured(url: String): ExtractResult {
        return extract(ExtractOptions(url = url, type = ExtractType.structured))
    }
    
    /**
     * Extract plain text from a webpage.
     */
    suspend fun extractText(url: String): ExtractResult {
        return extract(ExtractOptions(url = url, type = ExtractType.text))
    }
    
    /**
     * Extract all links from a webpage.
     */
    suspend fun extractLinks(url: String): ExtractResult {
        return extract(ExtractOptions(url = url, type = ExtractType.links))
    }
    
    /**
     * Extract all images from a webpage.
     */
    suspend fun extractImages(url: String): ExtractResult {
        return extract(ExtractOptions(url = url, type = ExtractType.images))
    }
    
    /**
     * Extract page metadata from a webpage.
     */
    suspend fun extractMetadata(url: String): ExtractResult {
        return extract(ExtractOptions(url = url, type = ExtractType.metadata))
    }

    private suspend fun <T> doRequest(method: String, path: String, body: T?): ByteArray {
        return withContext(Dispatchers.IO) {
            val url = URL("$baseUrl$path")
            val connection = url.openConnection() as HttpURLConnection

            try {
                connection.requestMethod = method
                connection.connectTimeout = timeout
                connection.readTimeout = timeout
                connection.setRequestProperty("X-Api-Key", apiKey)
                connection.setRequestProperty("Content-Type", "application/json")
                connection.setRequestProperty("User-Agent", USER_AGENT)

                if (body != null) {
                    connection.doOutput = true
                    val jsonBody = json.encodeToString(body)
                    connection.outputStream.use { it.write(jsonBody.toByteArray()) }
                }

                val responseCode = connection.responseCode

                if (responseCode >= 400) {
                    val errorBody = connection.errorStream?.use { it.readBytes() }
                        ?: ByteArray(0)
                    handleError(errorBody, responseCode)
                }

                connection.inputStream.use { it.readBytes() }
            } finally {
                connection.disconnect()
            }
        }
    }

    /**
     * Parse and throw a SnapAPIException from an HTTP error.
     */
    private fun handleError(body: ByteArray, statusCode: Int): Nothing {
        try {
            val errorResponse: ErrorResponse = json.decodeFromString(body.decodeToString())
            throw SnapAPIException(
                message = errorResponse.error.message,
                code = errorResponse.error.code,
                statusCode = statusCode,
                details = errorResponse.error.details
            )
        } catch (e: Exception) {
            if (e is SnapAPIException) throw e
            throw SnapAPIException(
                message = "HTTP $statusCode",
                code = "HTTP_ERROR",
                statusCode = statusCode
            )
        }
    }
}

/**
 * Exception thrown by SnapAPI operations.
 */
class SnapAPIException(
    message: String,
    val code: String,
    val statusCode: Int,
    val details: Map<String, kotlinx.serialization.json.JsonElement>? = null
) : Exception("[$code] $message (HTTP $statusCode)") {

    /**
     * Check if the error is retryable.
     */
    fun isRetryable(): Boolean {
        return code == "RATE_LIMITED" || code == "TIMEOUT" || statusCode >= 500
    }
}

/**
 * Extension function to decode base64 string to ByteArray.
 */
fun String.decodeBase64(): ByteArray = Base64.getDecoder().decode(this)

/**
 * Extension function to encode ByteArray to base64 string.
 */
fun ByteArray.encodeBase64(): String = Base64.getEncoder().encodeToString(this)
